### PR TITLE
Add/remove event handlers based on user setting

### PR DESCRIPTION
**Pull request by Pessimistress**
This is to fix https://github.com/uber/react-map-gl/issues/307
*Problem*
`EventManager.emit` is async, and therefore `event.preventDefault` may not be effective in handlers.
When scrollZoom is disabled, we need to unregister the event instead of filter it out in the handler.

*Changes*
- Disable recognizers in `EventManager.off` if no more handlers are attached
- Let `MapControls` manage its own events by passing in the `EventManager` instance

### CODE REVIEW

File: src/utils/event-manager/constants.js [hunk 1]
```diff
@@ -33,7 +33,6 @@ export const BASIC_EVENT_ALIASES = {
  * this block maps event names to the Recognizers required to detect the events.
  */
 export const EVENT_RECOGNIZER_MAP = {
-  click: 'tap',
```
Comment (ericsoco) [inline src/utils/event-manager/constants.js]: Why did you remove the `click` mapping?

File: src/utils/event-manager/move-input.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 const MOUSE_EVENTS = ['mousedown', 'mousemove', 'mouseup'];
-const MOVE_EVENT_TYPES = ['mousemove', 'pointermove'];
+const EVENT_TYPE = 'pointermove';
```
Comment (ericsoco) [inline src/utils/event-manager/move-input.js]: Is this basically a continuation of the work we did in EventManager to move everything over to only `pointer*` events?

File: src/utils/event-manager/move-input.js [hunk 3]
```diff
@@ -35,9 +35,9 @@ export default class MoveInput {
    * Enable this input (begin processing events)
    * if the specified event type is among those handled by this input.
    */
-  enableIfEventSupported(eventType) {
-    if (MOVE_EVENT_TYPES.indexOf(eventType) >= 0) {
-      this.options.enable = true;
+  enableIfEventSupported(eventType, enabled) {
```
Comment (ericsoco) [inline src/utils/event-manager/move-input.js]: Should probably change this to `setEnabledIfEventSupported`.

File: src/utils/event-manager/wheel-input.js [hunk 4]
```diff
@@ -64,6 +64,7 @@ export default class WheelInput {
     if (!this.options.enable) {
       return;
     }
+    event.preventDefault();
```
Comment (ericsoco) [inline src/utils/event-manager/wheel-input.js]: Didn't we intentionally remove this previously, for some reason?

Comment (ericsoco) [inline src/utils/event-manager/wheel-input.js]: Yeah, [here](https://github.com/uber/react-map-gl/pull/283).
> Fix bug where page scroll is blocked even with scrollZoom: false

Oh, is the idea that instead of blocking it, we're now conditionally adding the handler only when `scrollZoom` is `true`?

File: src/utils/map-controls.js [hunk 5]
```diff
@@ -25,36 +25,32 @@ const PITCH_MOUSE_THRESHOLD = 5;
...
-  'panmove',
-  'panend',
-  'pinchstart',
-  'pinch',
-  'pinchend',
-  'doubletap',
-  'wheel'
-];
+const EVENT_TYPES = {
+  wheel: ['wheel'],
```
Comment (ericsoco) [inline src/utils/map-controls.js]: These keys should probably be ALL_CAPS.

File: src/utils/event-manager/constants.js (hunk 1 above)
Comment (Pessimistress) [inline src/utils/event-manager/constants.js]: I changed the recognizer look up to use the aliased event name (click -> tap) so this entry is no longer necessary.

File: src/utils/event-manager/move-input.js (hunk 2 above)
Comment (Pessimistress) [inline src/utils/event-manager/move-input.js]: `mousemove` is aliased to `pointermove` so this extra event is never used.

File: src/utils/event-manager/wheel-input.js (hunk 4 above)
Comment (Pessimistress) [inline src/utils/event-manager/wheel-input.js]: Calling `event.preventDefault` from the event handler for `wheel` event does not stop the default of the trackpad scroll gesture. There's a timing issue related to how event manager emits events.

Comment (ericsoco) [inline src/utils/event-manager/wheel-input.js]: Ok, got it. It's just wheel events that are async, correct? (Due to the `setTimeout` in there.)

File: src/components/interactive-map.js [hunk 6]
```diff
@@ -151,20 +149,24 @@ export default class InteractiveMap extends PureComponent {
...
   componentDidMount() {
     const {eventCanvas} = this.refs;
-    const {mapControls} = this.props;
-
-    // Register event handlers defined by map controls
-    const events = {};
-    mapControls.events.forEach(eventName => {
-      events[eventName] = this._handleEvent;
-    });
+    const mapControls = this.props.mapControls || new MapControls();
```
Comment (abmai) [inline src/components/interactive-map.js]: Is this to explicitly guarding against `null`?

File: src/utils/event-manager/event-manager.js [hunk 7]
```diff
@@ -107,27 +107,32 @@ export default class EventManager {
     }
   }
 
+  /*
+   * Enable/disable recognizer for the given event
+   */
+  _enableRecognizer(name, enabled) {
```
Comment (abmai) [inline src/utils/event-manager/event-manager.js]: nit: `_toggleRecognizer` might be a better name since this is used for both enabling/disabling

File: src/utils/map-controls.js (hunk 5 above)
Comment (abmai) [inline src/utils/map-controls.js]: +1

File: src/utils/map-controls.js [hunk 8]
```diff
@@ -116,27 +112,55 @@ export default class MapControls {
...
+
+    // Interaction toggles
     this.scrollZoom = scrollZoom;
     this.dragPan = dragPan;
     this.dragRotate = dragRotate;
     this.doubleClickZoom = doubleClickZoom;
     this.touchZoomRotate = touchZoomRotate;
   }
 
+  enableEvents(eventNames, enabled) {
```
Comment (abmai) [inline src/utils/map-controls.js]: nit: `toggleEvents` instead?

File: src/components/interactive-map.js (hunk 6 above)
Comment (Pessimistress) [inline src/components/interactive-map.js]: We used to set a default value for the prop `mapControls`. The change introduced by this PR makes the state of `mapControls` dependent on `eventManager` - when the component mounts, a new `eventManager` is created, and `mapControls` must be reset.

This new behavior is potentially introducing bugs if the user is not managing their custom `mapControls` instance properly. We used to call this prop `MapControls` which expects a constructor. Any suggestions?

File: src/utils/event-manager/wheel-input.js (hunk 4 above)
Comment (Pessimistress) [inline src/utils/event-manager/wheel-input.js]: Seems that way. The `setTimeout` code originates from v1 and I don't really know its purpose.
